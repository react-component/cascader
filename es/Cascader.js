import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _toConsumableArray from 'babel-runtime/helpers/toConsumableArray';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Component, cloneElement } from 'react';
import PropTypes from 'prop-types';
import Trigger from 'rc-trigger';
import warning from 'warning';
import KeyCode from 'rc-util/es/KeyCode';
import arrayTreeFilter from 'array-tree-filter';
import shallowEqualArrays from 'shallow-equal/arrays';
import { polyfill } from 'react-lifecycles-compat';
import Menus from './Menus';

var BUILT_IN_PLACEMENTS = {
  bottomLeft: {
    points: ['tl', 'bl'],
    offset: [0, 4],
    overflow: {
      adjustX: 1,
      adjustY: 1,
    },
  },
  topLeft: {
    points: ['bl', 'tl'],
    offset: [0, -4],
    overflow: {
      adjustX: 1,
      adjustY: 1,
    },
  },
  bottomRight: {
    points: ['tr', 'br'],
    offset: [0, 4],
    overflow: {
      adjustX: 1,
      adjustY: 1,
    },
  },
  topRight: {
    points: ['br', 'tr'],
    offset: [0, -4],
    overflow: {
      adjustX: 1,
      adjustY: 1,
    },
  },
};

var Cascader = (function(_Component) {
  _inherits(Cascader, _Component);

  function Cascader(props) {
    _classCallCheck(this, Cascader);

    var _this = _possibleConstructorReturn(
      this,
      (Cascader.__proto__ || Object.getPrototypeOf(Cascader)).call(this, props),
    );

    _this.setPopupVisible = function(popupVisible) {
      if (!('popupVisible' in _this.props)) {
        _this.setState({ popupVisible: popupVisible });
      }
      // sync activeValue with value when panel open
      if (popupVisible && !_this.state.popupVisible) {
        _this.setState({
          activeValue: _this.state.value,
        });
      }
      _this.props.onPopupVisibleChange(popupVisible);
    };

    _this.handleChange = function(options, setProps, e) {
      if (e.type !== 'keydown' || e.keyCode === KeyCode.ENTER) {
        _this.props.onChange(
          options.map(function(o) {
            return o[_this.getFieldName('value')];
          }),
          options,
        );
        _this.setPopupVisible(setProps.visible);
      }
    };

    _this.handlePopupVisibleChange = function(popupVisible) {
      _this.setPopupVisible(popupVisible);
    };

    _this.handleMenuSelect = function(targetOption, menuIndex, e) {
      // Keep focused state for keyboard support
      var triggerNode = _this.trigger.getRootDomNode();
      if (triggerNode && triggerNode.focus) {
        triggerNode.focus();
      }
      var _this$props = _this.props,
        changeOnSelect = _this$props.changeOnSelect,
        loadData = _this$props.loadData,
        expandTrigger = _this$props.expandTrigger;

      if (!targetOption || targetOption.disabled) {
        return;
      }
      var activeValue = _this.state.activeValue;

      activeValue = activeValue.slice(0, menuIndex + 1);
      activeValue[menuIndex] = targetOption[_this.getFieldName('value')];
      var activeOptions = _this.getActiveOptions(activeValue);
      if (
        targetOption.isLeaf === false &&
        !targetOption[_this.getFieldName('children')] &&
        loadData
      ) {
        if (changeOnSelect) {
          _this.handleChange(activeOptions, { visible: true }, e);
        }
        _this.setState({ activeValue: activeValue });
        loadData(activeOptions);
        return;
      }
      var newState = {};
      if (
        !targetOption[_this.getFieldName('children')] ||
        !targetOption[_this.getFieldName('children')].length
      ) {
        _this.handleChange(activeOptions, { visible: false }, e);
        // set value to activeValue when select leaf option
        newState.value = activeValue;
        // add e.type judgement to prevent `onChange` being triggered by mouseEnter
      } else if (changeOnSelect && (e.type === 'click' || e.type === 'keydown')) {
        if (expandTrigger === 'hover') {
          _this.handleChange(activeOptions, { visible: false }, e);
        } else {
          _this.handleChange(activeOptions, { visible: true }, e);
        }
        // set value to activeValue on every select
        newState.value = activeValue;
      }
      newState.activeValue = activeValue;
      //  not change the value by keyboard
      if ('value' in _this.props || (e.type === 'keydown' && e.keyCode !== KeyCode.ENTER)) {
        delete newState.value;
      }
      _this.setState(newState);
    };

    _this.handleItemDoubleClick = function() {
      var changeOnSelect = _this.props.changeOnSelect;

      if (changeOnSelect) {
        _this.setPopupVisible(false);
      }
    };

    _this.handleKeyDown = function(e) {
      var children = _this.props.children;
      // https://github.com/ant-design/ant-design/issues/6717
      // Don't bind keyboard support when children specify the onKeyDown

      if (children && children.props.onKeyDown) {
        children.props.onKeyDown(e);
        return;
      }
      var activeValue = [].concat(_toConsumableArray(_this.state.activeValue));
      var currentLevel = activeValue.length - 1 < 0 ? 0 : activeValue.length - 1;
      var currentOptions = _this.getCurrentLevelOptions();
      var currentIndex = currentOptions
        .map(function(o) {
          return o[_this.getFieldName('value')];
        })
        .indexOf(activeValue[currentLevel]);
      if (
        e.keyCode !== KeyCode.DOWN &&
        e.keyCode !== KeyCode.UP &&
        e.keyCode !== KeyCode.LEFT &&
        e.keyCode !== KeyCode.RIGHT &&
        e.keyCode !== KeyCode.ENTER &&
        e.keyCode !== KeyCode.BACKSPACE &&
        e.keyCode !== KeyCode.ESC
      ) {
        return;
      }
      // Press any keys above to reopen menu
      if (
        !_this.state.popupVisible &&
        e.keyCode !== KeyCode.BACKSPACE &&
        e.keyCode !== KeyCode.LEFT &&
        e.keyCode !== KeyCode.RIGHT &&
        e.keyCode !== KeyCode.ESC
      ) {
        _this.setPopupVisible(true);
        return;
      }
      if (e.keyCode === KeyCode.DOWN || e.keyCode === KeyCode.UP) {
        e.preventDefault();
        var nextIndex = currentIndex;
        if (nextIndex !== -1) {
          if (e.keyCode === KeyCode.DOWN) {
            nextIndex += 1;
            nextIndex = nextIndex >= currentOptions.length ? 0 : nextIndex;
          } else {
            nextIndex -= 1;
            nextIndex = nextIndex < 0 ? currentOptions.length - 1 : nextIndex;
          }
        } else {
          nextIndex = 0;
        }
        activeValue[currentLevel] = currentOptions[nextIndex][_this.getFieldName('value')];
      } else if (e.keyCode === KeyCode.LEFT || e.keyCode === KeyCode.BACKSPACE) {
        e.preventDefault();
        activeValue.splice(activeValue.length - 1, 1);
      } else if (e.keyCode === KeyCode.RIGHT) {
        e.preventDefault();
        if (
          currentOptions[currentIndex] &&
          currentOptions[currentIndex][_this.getFieldName('children')]
        ) {
          activeValue.push(
            currentOptions[currentIndex][_this.getFieldName('children')][0][
              _this.getFieldName('value')
            ],
          );
        }
      } else if (e.keyCode === KeyCode.ESC) {
        _this.setPopupVisible(false);
        return;
      }
      if (!activeValue || activeValue.length === 0) {
        _this.setPopupVisible(false);
      }
      var activeOptions = _this.getActiveOptions(activeValue);
      var targetOption = activeOptions[activeOptions.length - 1];
      _this.handleMenuSelect(targetOption, activeOptions.length - 1, e);

      if (_this.props.onKeyDown) {
        _this.props.onKeyDown(e);
      }
    };

    _this.saveTrigger = function(node) {
      _this.trigger = node;
    };

    var initialValue = [];
    if ('value' in props) {
      initialValue = props.value || [];
    } else if ('defaultValue' in props) {
      initialValue = props.defaultValue || [];
    }

    warning(
      !('filedNames' in props),
      '`filedNames` of Cascader is a typo usage and deprecated, please use `fieldNames` instead.',
    );

    _this.state = {
      popupVisible: props.popupVisible,
      activeValue: initialValue,
      value: initialValue,
      prevProps: props,
    };
    _this.defaultFieldNames = { label: 'label', value: 'value', children: 'children' };
    return _this;
  }

  _createClass(
    Cascader,
    [
      {
        key: 'getPopupDOMNode',
        value: function getPopupDOMNode() {
          return this.trigger.getPopupDomNode();
        },
      },
      {
        key: 'getFieldName',
        value: function getFieldName(name) {
          var defaultFieldNames = this.defaultFieldNames;
          var _props = this.props,
            fieldNames = _props.fieldNames,
            filedNames = _props.filedNames;

          if ('filedNames' in this.props) {
            return filedNames[name] || defaultFieldNames[name]; // For old compatibility
          }
          return fieldNames[name] || defaultFieldNames[name];
        },
      },
      {
        key: 'getFieldNames',
        value: function getFieldNames() {
          var _props2 = this.props,
            fieldNames = _props2.fieldNames,
            filedNames = _props2.filedNames;

          if ('filedNames' in this.props) {
            return filedNames; // For old compatibility
          }
          return fieldNames;
        },
      },
      {
        key: 'getCurrentLevelOptions',
        value: function getCurrentLevelOptions() {
          var _this2 = this;

          var _props$options = this.props.options,
            options = _props$options === undefined ? [] : _props$options;
          var _state$activeValue = this.state.activeValue,
            activeValue = _state$activeValue === undefined ? [] : _state$activeValue;

          var result = arrayTreeFilter(
            options,
            function(o, level) {
              return o[_this2.getFieldName('value')] === activeValue[level];
            },
            { childrenKeyName: this.getFieldName('children') },
          );
          if (result[result.length - 2]) {
            return result[result.length - 2][this.getFieldName('children')];
          }
          return [].concat(_toConsumableArray(options)).filter(function(o) {
            return !o.disabled;
          });
        },
      },
      {
        key: 'getActiveOptions',
        value: function getActiveOptions(activeValue) {
          var _this3 = this;

          return arrayTreeFilter(
            this.props.options || [],
            function(o, level) {
              return o[_this3.getFieldName('value')] === activeValue[level];
            },
            { childrenKeyName: this.getFieldName('children') },
          );
        },
      },
      {
        key: 'render',
        value: function render() {
          var _props3 = this.props,
            prefixCls = _props3.prefixCls,
            transitionName = _props3.transitionName,
            popupClassName = _props3.popupClassName,
            _props3$options = _props3.options,
            options = _props3$options === undefined ? [] : _props3$options,
            disabled = _props3.disabled,
            builtinPlacements = _props3.builtinPlacements,
            popupPlacement = _props3.popupPlacement,
            children = _props3.children,
            restProps = _objectWithoutProperties(_props3, [
              'prefixCls',
              'transitionName',
              'popupClassName',
              'options',
              'disabled',
              'builtinPlacements',
              'popupPlacement',
              'children',
            ]);
          // Did not show popup when there is no options

          var menus = React.createElement('div', null);
          var emptyMenuClassName = '';
          if (options && options.length > 0) {
            menus = React.createElement(
              Menus,
              _extends({}, this.props, {
                fieldNames: this.getFieldNames(),
                defaultFieldNames: this.defaultFieldNames,
                activeValue: this.state.activeValue,
                onSelect: this.handleMenuSelect,
                onItemDoubleClick: this.handleItemDoubleClick,
                visible: this.state.popupVisible,
              }),
            );
          } else {
            emptyMenuClassName = ' ' + prefixCls + '-menus-empty';
          }
          return React.createElement(
            Trigger,
            _extends(
              {
                ref: this.saveTrigger,
              },
              restProps,
              {
                options: options,
                disabled: disabled,
                popupPlacement: popupPlacement,
                builtinPlacements: builtinPlacements,
                popupTransitionName: transitionName,
                action: disabled ? [] : ['click'],
                popupVisible: disabled ? false : this.state.popupVisible,
                onPopupVisibleChange: this.handlePopupVisibleChange,
                prefixCls: prefixCls + '-menus',
                popupClassName: popupClassName + emptyMenuClassName,
                popup: menus,
              },
            ),
            cloneElement(children, {
              onKeyDown: this.handleKeyDown,
              tabIndex: disabled ? undefined : 0,
            }),
          );
        },
      },
    ],
    [
      {
        key: 'getDerivedStateFromProps',
        value: function getDerivedStateFromProps(nextProps, prevState) {
          var _prevState$prevProps = prevState.prevProps,
            prevProps = _prevState$prevProps === undefined ? {} : _prevState$prevProps;

          var newState = {
            prevProps: nextProps,
          };

          if ('value' in nextProps && !shallowEqualArrays(prevProps.value, nextProps.value)) {
            newState.value = nextProps.value || [];

            // allow activeValue diff from value
            // https://github.com/ant-design/ant-design/issues/2767
            if (!('loadData' in nextProps)) {
              newState.activeValue = nextProps.value || [];
            }
          }
          if ('popupVisible' in nextProps) {
            newState.popupVisible = nextProps.popupVisible;
          }

          return newState;
        },
      },
    ],
  );

  return Cascader;
})(Component);

Cascader.defaultProps = {
  onChange: function onChange() {},
  onPopupVisibleChange: function onPopupVisibleChange() {},

  disabled: false,
  transitionName: '',
  prefixCls: 'rc-cascader',
  popupClassName: '',
  popupPlacement: 'bottomLeft',
  builtinPlacements: BUILT_IN_PLACEMENTS,
  expandTrigger: 'click',
  fieldNames: { label: 'label', value: 'value', children: 'children' },
  expandIcon: '>',
};

Cascader.propTypes = {
  value: PropTypes.array,
  defaultValue: PropTypes.array,
  options: PropTypes.array.isRequired,
  onChange: PropTypes.func,
  onPopupVisibleChange: PropTypes.func,
  popupVisible: PropTypes.bool,
  disabled: PropTypes.bool,
  transitionName: PropTypes.string,
  popupClassName: PropTypes.string,
  popupPlacement: PropTypes.string,
  prefixCls: PropTypes.string,
  dropdownMenuColumnStyle: PropTypes.object,
  builtinPlacements: PropTypes.object,
  loadData: PropTypes.func,
  changeOnSelect: PropTypes.bool,
  children: PropTypes.node,
  onKeyDown: PropTypes.func,
  expandTrigger: PropTypes.string,
  fieldNames: PropTypes.object,
  filedNames: PropTypes.object, // typo but for compatibility
  expandIcon: PropTypes.node,
  loadingIcon: PropTypes.node,
};

polyfill(Cascader);

export default Cascader;
