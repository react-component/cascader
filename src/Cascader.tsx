import * as React from 'react';
import Trigger, { BuildInPlacements, TriggerProps } from 'rc-trigger';
import warning from 'warning';
import KeyCode from 'rc-util/lib/KeyCode';
import arrayTreeFilter from 'array-tree-filter';
import { isEqualArrays } from './utils';
import Menus from './Menus';
import BUILT_IN_PLACEMENTS from './placements';

export interface CascaderFieldNames {
  value?: string;
  label?: string;
  children?: string;
}

export interface CascaderOption {
  value?: string;
  label?: React.ReactNode;
  disabled?: boolean;
  isLeaf?: boolean;
  loading?: boolean;
  children?: CascaderOption[];
  [key: string]: any;
}

export interface CascaderProps extends Pick<TriggerProps, 'getPopupContainer'> {
  value?: string[];
  defaultValue?: string[];
  options?: CascaderOption[];
  onChange?: (value: string[], selectOptions: CascaderOption[]) => void;
  onPopupVisibleChange?: (popupVisible: boolean) => void;
  popupVisible?: boolean;
  disabled?: boolean;
  transitionName?: string;
  popupClassName?: string;
  popupPlacement?: string;
  prefixCls?: string;
  dropdownMenuColumnStyle?: React.CSSProperties;
  builtinPlacements?: BuildInPlacements;
  loadData?: (selectOptions: CascaderOption[]) => void;
  changeOnSelect?: boolean;
  children?: React.ReactElement;
  onKeyDown?: (e: React.KeyboardEvent<HTMLElement>) => void;
  expandTrigger?: string;
  fieldNames?: CascaderFieldNames;
  filedNames?: CascaderFieldNames; // typo but for compatibility
  expandIcon?: React.ReactNode;
  loadingIcon?: React.ReactNode;
}

interface CascaderState {
  popupVisible?: boolean;
  activeValue?: string[];
  value?: string[];
  prevProps?: CascaderProps;
}

class Cascader extends React.Component<CascaderProps, CascaderState> {
  defaultFieldNames: object;

  trigger: any;

  constructor(props: CascaderProps) {
    super(props);
    let initialValue = [];
    if ('value' in props) {
      initialValue = props.value || [];
    } else if ('defaultValue' in props) {
      initialValue = props.defaultValue || [];
    }

    warning(
      !('filedNames' in props),
      '`filedNames` of Cascader is a typo usage and deprecated, please use `fieldNames` instead.',
    );

    this.state = {
      popupVisible: props.popupVisible,
      activeValue: initialValue,
      value: initialValue,
      prevProps: props,
    };
    this.defaultFieldNames = {
      label: 'label',
      value: 'value',
      children: 'children',
    };
  }

  static defaultProps: CascaderProps = {
    onChange: () => {},
    onPopupVisibleChange: () => {},
    disabled: false,
    transitionName: '',
    prefixCls: 'rc-cascader',
    popupClassName: '',
    popupPlacement: 'bottomLeft',
    builtinPlacements: BUILT_IN_PLACEMENTS,
    expandTrigger: 'click',
    fieldNames: { label: 'label', value: 'value', children: 'children' },
    expandIcon: '>',
  };

  static getDerivedStateFromProps(nextProps: CascaderProps, prevState: CascaderState) {
    const { prevProps = {} } = prevState;
    const newState: CascaderState = {
      prevProps: nextProps,
    };

    if ('value' in nextProps && !isEqualArrays(prevProps.value, nextProps.value)) {
      newState.value = nextProps.value || [];

      // allow activeValue diff from value
      // https://github.com/ant-design/ant-design/issues/2767
      if (!('loadData' in nextProps)) {
        newState.activeValue = nextProps.value || [];
      }
    }
    if ('popupVisible' in nextProps) {
      newState.popupVisible = nextProps.popupVisible;
    }

    return newState;
  }

  getPopupDOMNode() {
    return this.trigger.getPopupDomNode();
  }

  getFieldName(name: string): string {
    const { defaultFieldNames } = this;
    const { fieldNames, filedNames } = this.props;
    if ('filedNames' in this.props) {
      return filedNames[name] || defaultFieldNames[name]; // For old compatibility
    }
    return fieldNames[name] || defaultFieldNames[name];
  }

  getFieldNames(): CascaderFieldNames {
    const { fieldNames, filedNames } = this.props;
    if ('filedNames' in this.props) {
      return filedNames; // For old compatibility
    }
    return fieldNames;
  }

  getCurrentLevelOptions(): CascaderOption[] {
    const { options = [] } = this.props;
    const { activeValue = [] } = this.state;
    const result = arrayTreeFilter(
      options,
      (o, level) => o[this.getFieldName('value')] === activeValue[level],
      { childrenKeyName: this.getFieldName('children') },
    );
    if (result[result.length - 2]) {
      return result[result.length - 2][this.getFieldName('children')];
    }
    return [...options].filter(o => !o.disabled);
  }

  getActiveOptions(activeValue: string[]): CascaderOption[] {
    return arrayTreeFilter(
      this.props.options || [],
      (o, level) => o[this.getFieldName('value')] === activeValue[level],
      { childrenKeyName: this.getFieldName('children') },
    );
  }

  setPopupVisible = (popupVisible: boolean) => {
    const { value } = this.state;
    if (!('popupVisible' in this.props)) {
      this.setState({ popupVisible });
    }
    // sync activeValue with value when panel open
    if (popupVisible && !this.state.popupVisible) {
      this.setState({
        activeValue: value,
      });
    }
    this.props.onPopupVisibleChange(popupVisible);
  };

  handleChange = (options: CascaderOption[], { visible }, e: React.KeyboardEvent<HTMLElement>) => {
    if (e.type !== 'keydown' || e.keyCode === KeyCode.ENTER) {
      this.props.onChange(options.map(o => o[this.getFieldName('value')]), options);
      this.setPopupVisible(visible);
    }
  };

  handlePopupVisibleChange = (popupVisible: boolean) => {
    this.setPopupVisible(popupVisible);
  };

  handleMenuSelect = (
    targetOption: CascaderOption,
    menuIndex: number,
    e: React.KeyboardEvent<HTMLElement>,
  ) => {
    // Keep focused state for keyboard support
    const triggerNode = this.trigger.getRootDomNode();
    if (triggerNode && triggerNode.focus) {
      triggerNode.focus();
    }
    const { changeOnSelect, loadData, expandTrigger } = this.props;
    if (!targetOption || targetOption.disabled) {
      return;
    }
    let { activeValue } = this.state;
    activeValue = activeValue.slice(0, menuIndex + 1);
    activeValue[menuIndex] = targetOption[this.getFieldName('value')];
    const activeOptions = this.getActiveOptions(activeValue);
    if (targetOption.isLeaf === false && !targetOption[this.getFieldName('children')] && loadData) {
      if (changeOnSelect) {
        this.handleChange(activeOptions, { visible: true }, e);
      }
      this.setState({ activeValue });
      loadData(activeOptions);
      return;
    }
    const newState: CascaderState = {};
    if (
      !targetOption[this.getFieldName('children')] ||
      !targetOption[this.getFieldName('children')].length
    ) {
      this.handleChange(activeOptions, { visible: false }, e);
      // set value to activeValue when select leaf option
      newState.value = activeValue;
      // add e.type judgement to prevent `onChange` being triggered by mouseEnter
    } else if (changeOnSelect && (e.type === 'click' || e.type === 'keydown')) {
      if (expandTrigger === 'hover') {
        this.handleChange(activeOptions, { visible: false }, e);
      } else {
        this.handleChange(activeOptions, { visible: true }, e);
      }
      // set value to activeValue on every select
      newState.value = activeValue;
    }
    newState.activeValue = activeValue;
    //  not change the value by keyboard
    if ('value' in this.props || (e.type === 'keydown' && e.keyCode !== KeyCode.ENTER)) {
      delete newState.value;
    }
    this.setState(newState);
  };

  handleItemDoubleClick = () => {
    const { changeOnSelect } = this.props;
    if (changeOnSelect) {
      this.setPopupVisible(false);
    }
  };

  handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    const { children } = this.props;
    // https://github.com/ant-design/ant-design/issues/6717
    // Don't bind keyboard support when children specify the onKeyDown
    if (children && children.props.onKeyDown) {
      children.props.onKeyDown(e);
      return;
    }
    const activeValue = [...this.state.activeValue];
    const currentLevel = activeValue.length - 1 < 0 ? 0 : activeValue.length - 1;
    const currentOptions = this.getCurrentLevelOptions();
    const currentIndex = currentOptions
      .map(o => o[this.getFieldName('value')])
      .indexOf(activeValue[currentLevel]);
    if (
      e.keyCode !== KeyCode.DOWN &&
      e.keyCode !== KeyCode.UP &&
      e.keyCode !== KeyCode.LEFT &&
      e.keyCode !== KeyCode.RIGHT &&
      e.keyCode !== KeyCode.ENTER &&
      e.keyCode !== KeyCode.SPACE &&
      e.keyCode !== KeyCode.BACKSPACE &&
      e.keyCode !== KeyCode.ESC &&
      e.keyCode !== KeyCode.TAB
    ) {
      return;
    }
    // Press any keys above to reopen menu
    if (
      !this.state.popupVisible &&
      e.keyCode !== KeyCode.BACKSPACE &&
      e.keyCode !== KeyCode.LEFT &&
      e.keyCode !== KeyCode.RIGHT &&
      e.keyCode !== KeyCode.ESC &&
      e.keyCode !== KeyCode.TAB
    ) {
      this.setPopupVisible(true);
      return;
    }
    if (e.keyCode === KeyCode.DOWN || e.keyCode === KeyCode.UP) {
      e.preventDefault();
      let nextIndex = currentIndex;
      if (nextIndex !== -1) {
        if (e.keyCode === KeyCode.DOWN) {
          nextIndex += 1;
          nextIndex = nextIndex >= currentOptions.length ? 0 : nextIndex;
        } else {
          nextIndex -= 1;
          nextIndex = nextIndex < 0 ? currentOptions.length - 1 : nextIndex;
        }
      } else {
        nextIndex = 0;
      }
      activeValue[currentLevel] = currentOptions[nextIndex][this.getFieldName('value')];
    } else if (e.keyCode === KeyCode.LEFT || e.keyCode === KeyCode.BACKSPACE) {
      e.preventDefault();
      activeValue.splice(activeValue.length - 1, 1);
    } else if (e.keyCode === KeyCode.RIGHT) {
      e.preventDefault();
      if (
        currentOptions[currentIndex] &&
        currentOptions[currentIndex][this.getFieldName('children')]
      ) {
        activeValue.push(
          currentOptions[currentIndex][this.getFieldName('children')][0][
            this.getFieldName('value')
          ],
        );
      }
    } else if (e.keyCode === KeyCode.ESC || e.keyCode === KeyCode.TAB) {
      this.setPopupVisible(false);
      return;
    }
    if (!activeValue || activeValue.length === 0) {
      this.setPopupVisible(false);
    }
    const activeOptions = this.getActiveOptions(activeValue);
    const targetOption = activeOptions[activeOptions.length - 1];
    this.handleMenuSelect(targetOption, activeOptions.length - 1, e);

    if (this.props.onKeyDown) {
      this.props.onKeyDown(e);
    }
  };

  saveTrigger = node => {
    this.trigger = node;
  };

  render() {
    const {
      prefixCls,
      transitionName,
      popupClassName,
      options = [],
      disabled,
      builtinPlacements,
      popupPlacement,
      children,
      ...restProps
    } = this.props;
    // Did not show popup when there is no options
    let menus = <div />;
    let emptyMenuClassName = '';
    if (options && options.length > 0) {
      menus = (
        <Menus
          {...this.props}
          fieldNames={this.getFieldNames()}
          defaultFieldNames={this.defaultFieldNames}
          activeValue={this.state.activeValue}
          onSelect={this.handleMenuSelect}
          onItemDoubleClick={this.handleItemDoubleClick}
          visible={this.state.popupVisible}
        />
      );
    } else {
      emptyMenuClassName = ` ${prefixCls}-menus-empty`;
    }
    return (
      <Trigger
        ref={this.saveTrigger}
        {...restProps}
        popupPlacement={popupPlacement}
        builtinPlacements={builtinPlacements}
        popupTransitionName={transitionName}
        action={disabled ? [] : ['click']}
        popupVisible={disabled ? false : this.state.popupVisible}
        onPopupVisibleChange={this.handlePopupVisibleChange}
        prefixCls={`${prefixCls}-menus`}
        popupClassName={popupClassName + emptyMenuClassName}
        popup={menus}
      >
        {React.cloneElement(children, {
          onKeyDown: this.handleKeyDown,
          tabIndex: disabled ? undefined : 0,
        })}
      </Trigger>
    );
  }
}

export default Cascader;
