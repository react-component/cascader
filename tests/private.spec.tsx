/* eslint-disable @typescript-eslint/consistent-type-imports */

import React from 'react';
import { mount } from './enzyme';
import Cascader from '../src';

describe('Cascader.Private', () => {
  it('dropdownPrefixCls', () => {
    const wrapper = mount(
      <Cascader
        defaultValue={['light', 'toy']}
        options={[
          {
            label: 'Light',
            value: 'light',
            children: [
              {
                label: 'Toy',
                value: 'toy',
              },
            ],
          },
        ]}
        open
        prefixCls="bamboo"
        dropdownPrefixCls="little"
      />,
    );

    expect(wrapper.render()).toMatchSnapshot();
  });
});
