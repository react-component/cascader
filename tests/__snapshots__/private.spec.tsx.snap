// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cascader.Private dropdownPrefixCls 1`] = `
<div
  class="bamboo bamboo-single bamboo-show-arrow bamboo-open"
>
  <div
    class="bamboo-selector"
  >
    <span
      class="bamboo-selection-search"
    >
      <input
        aria-autocomplete="list"
        aria-controls="rc_select_TEST_OR_SSR_list"
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-owns="rc_select_TEST_OR_SSR_list"
        autocomplete="off"
        class="bamboo-selection-search-input"
        id="rc_select_TEST_OR_SSR"
        readonly=""
        role="combobox"
        style="opacity: 0;"
        type="search"
        unselectable="on"
        value=""
      />
    </span>
    <span
      class="bamboo-selection-item"
      title="Light / Toy"
    >
      Light / Toy
    </span>
  </div>
  <div>
    <div
      class="bamboo-dropdown"
      style="opacity: 0; pointer-events: none; min-width: auto;"
    >
      <div>
        <div
          class="little-menus"
        >
          <ul
            class="little-menu"
            role="menu"
          >
            <li
              aria-checked="false"
              class="little-menu-item little-menu-item-expand little-menu-item-active"
              data-path-key="light"
              role="menuitemcheckbox"
              title="Light"
            >
              <div
                class="little-menu-item-content"
              >
                Light
              </div>
              <div
                class="little-menu-item-expand-icon"
              >
                &gt;
              </div>
            </li>
          </ul>
          <ul
            class="little-menu"
            role="menu"
          >
            <li
              aria-checked="true"
              class="little-menu-item little-menu-item-active"
              data-path-key="light__RC_CASCADER_SPLIT__toy"
              role="menuitemcheckbox"
              title="Toy"
            >
              <div
                class="little-menu-item-content"
              >
                Toy
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <span
    aria-hidden="true"
    class="bamboo-arrow"
    style="user-select: none;"
    unselectable="on"
  >
    <span
      class="bamboo-arrow-icon"
    />
  </span>
</div>
`;
